com.unity.perception/Tests/Runtime/GroundTruthTests/DatasetCaptureSensorSchedulingTests.cs

```diff
@@ -7,16 +9,78 @@
 
 namespace GroundTruthTests
 {
+    // Provides accessors and invocation methods for members of SimulationState that would otherwise be in-accessible
+    // due to protection level - use only when testing protected logic is critical
+    class SimulationStateTestHelper
+    {
+        SimulationState m_State => DatasetCapture.SimulationState;
+        Dictionary<SensorHandle, SimulationState.SensorData> m_SensorsReference;
+        MethodInfo m_SequenceTimeOfNextCaptureMethod;
+
+        internal SimulationStateTestHelper()
+        {
+            var bindingFlags = BindingFlags.NonPublic | BindingFlags.Instance;
```

Comment:
Instead of reflection, it's cleaner to mark the members you need `internal`. Then the API will be available in your tests.

Reply:
I think it depends on your definition of "cleaner"  - less code? Sure.  But, changing protection levels to enable unit tests also feels... pretty dirty.  In a nominal case we'd want to be testing this math at the API level, rather than turning the math itself into part of the internal API. Given that we've decided this is worth testing, even as private method, though, I'd rather hack the unit test than hack the interface.

Reply:
Testing internal state is unfortunately fairly common, and not always is it possible to use reflection to do the testing you need. For example, you are already using the internal type SimulationState.SensorData in this API. It was marked public (in an internal class) originally so that it could be used in tests, and to do otherwise would be a much bigger pain than just calling a method or assigning a field. While it's nice to not change the internal API of a class, it doesn't scale well in general and increases maintenance cost of tests.

This isn't a huge deal at the end of the day so I'll mark this approved and you can do what you think is right.

Reply:
I'm going to push this, but I have a feeling we're going to have this conversation again, so I'm also going to list my justifications, as there are quite a few red flags in your line of thinking:
1. "it's common" is not justification for employing bad practices
2. SimulationState.SensorData being public is tech debt. Using existing tech debt to justify adding more tech debt leads to an unmaintainable codebase
3. I did not make the claim that this scales well, and in fact it's not meant to. Testing at this level of granularity implicitly does not scale well and is in general bad practice. Because timing code tends to be critical, I'm ok with adding this test,  but this is an exception to the rule. It _should_ be hard to scale up testing in ways that do not follow good practices, and we _should not_ use one-off exceptions like this to set precedent for future bad decisions.